src/repr/src/fixed_length.rs

```diff
@@ -15,43 +15,41 @@
 //! or fixed-length containers standing in for them.
 
 use std::borrow::Borrow;
-use std::iter;
-use std::mem::size_of;
+use std::iter::{empty, Empty};
 
-use chrono::{DateTime, Utc};
-use mz_persist_types::Codec64;
-use ordered_float::OrderedFloat;
-use serde::{Deserialize, Serialize};
-use serde_big_array::BigArray;
-use timely::container::columnation::{Columnation, CopyRegion};
-use uuid::Uuid;
-
-use crate::adt::interval::Interval;
-use crate::adt::mz_acl_item::{AclItem, MzAclItem};
-use crate::adt::timestamp::CheckedTimestamp;
-use crate::row::{
-    date_to_array, naive_date_to_arrays, read_byte_array, read_date, read_naive_date, read_time,
-    time_to_arrays,
-};
-use crate::{ColumnType, Datum, Row, ScalarType, Timestamp};
-
-/// A typedef to ease use of 9-byte fixed-length representations.
-pub type Bytes9 = BytesN<9>;
+use crate::row::DatumListIter;
+use crate::{ColumnType, Datum, Row};
 
 /// A helper trait to get references to `Row` based on type information that only manifests
 /// at runtime (typically originating from inferred schemas).
 pub trait IntoRowByTypes: Sized {
+    /// An iterator type for use in `into_datum_iter`.
+    type Iter<'a>
```

Comment:
Can simplify the trait bounds:
```
diff --git a/src/repr/src/fixed_length.rs b/src/repr/src/fixed_length.rs
index 4374b9ad5d..02236ca5e0 100644
--- a/src/repr/src/fixed_length.rs
+++ b/src/repr/src/fixed_length.rs
@@ -24,7 +24,7 @@ use crate::{ColumnType, Datum, Row};
 /// at runtime (typically originating from inferred schemas).
 pub trait IntoRowByTypes: Sized {
     /// An iterator type for use in `into_datum_iter`.
-    type Iter<'a>
+    type Iter<'a>: Iterator<Item = Datum<'a>>
     where
         Self: 'a;

@@ -40,9 +40,7 @@ pub trait IntoRowByTypes: Sized {
     /// by `types`.
     ///
     /// Implementations are free to place specific requirements on the given schema.
-    fn into_datum_iter<'a>(&'a self, types: Option<&[ColumnType]>) -> Self::Iter<'a>
-    where
-        Self::Iter<'a>: Iterator<Item = Datum<'a>>;
+    fn into_datum_iter<'a>(&'a self, types: Option<&[ColumnType]>) -> Self::Iter<'a>;
 }

 // Blanket identity implementation for Row.
@@ -65,12 +63,9 @@ impl IntoRowByTypes for Row {
     /// This implementation panics if `types` other than `None` are provided. This is because
     /// `Row` is already self-describing and can use variable-length types, so we are
     /// explicitly not validating the given schema.
-    fn into_datum_iter<'a>(&'a self, types: Option<&[ColumnType]>) -> Self::Iter<'a>
-    where
-        Self::Iter<'a>: Iterator<Item = Datum<'a>>,
-    {
+    fn into_datum_iter<'a>(&'a self, types: Option<&[ColumnType]>) -> Self::Iter<'a> {
         assert!(types.is_none());
-        (*self.borrow()).iter()
+        self.iter()
     }
 }

@@ -145,10 +140,7 @@ impl IntoRowByTypes for () {
     ///
     /// This implementation panics if `types` other than `Some(&[])` are provided. This is because
     /// unit values need to have an empty schema.
-    fn into_datum_iter<'a>(&'a self, types: Option<&[ColumnType]>) -> Self::Iter<'a>
-    where
-        Self::Iter<'a>: Iterator<Item = Datum<'a>>,
-    {
+    fn into_datum_iter<'a>(&'a self, types: Option<&[ColumnType]>) -> Self::Iter<'a> {
         let Some(&[]) = types else {
             panic!("Non-empty schema with unit values")
         };
```

Reply:
Cool, incorporated, thanks!